### PR TITLE
Duplicate regression tests in InkAPITest.cc as Au Tests

### PR DESCRIPTION
**Pull request by ywkaras**
This PR has now absorbed https://github.com/apache/trafficserver/pull/6489 .

### CODE REVIEW

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
This contributes to resolving https://github.com/apache/trafficserver/issues/6269 .

**Comment by ywkaras** [comment]
I'm going to close this and include these changes in https://github.com/apache/trafficserver/pull/6832 .

**Comment by ywkaras** [comment]
I'm going to close https://github.com/apache/trafficserver/pull/6489 and absorb the changes into this PR.

**Comment by ywkaras** [comment]
[approve ci autest]

File: tests/gold_tests/pluginTest/tsapi2/alt_info.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            log("request id number = %d", x_req_num);
+
+            auto data = static_cast<ContData *>(TSContDataGet(contp));
+            data->test(checkHttpTxnReqOrResp(log, info, TSHttpAltInfoCachedReqGet, "alt info cached request", 6));
+            data->test(
+              checkHttpTxnReqOrResp(log, info, TSHttpAltInfoCachedRespGet, "alt info cached response", 6, TS_HTTP_STATUS_OK));
+
+            if (7 == x_req_num) {
+              // This function does not actually seem to do anything.
+              TSHttpAltInfoQualitySet(info, 0.5);
```
Comment (ywkaras) [inline tests/gold_tests/pluginTest/tsapi2/alt_info.h:77]: This function also seems to have no effect in the regression tests that this test code is based on:
```
Breakpoint 1, synclient_txn_read_response (contp=<optimized out>) at traffic_server/InkAPITest.cc:669
669	  TSDebug(CDBG_TAG, "Response = |%s|, req len = %d", txn->response, txn->response_len);
(gdb) p txn->response
$3 = "HTTP/1.0 200 OK\r\nX-Response-ID: 6\r\nCache-Control: max-age=86400\r\nContent-Language: English\r\nDate: Tue, 08 Sep 2020 14:41:34 GMT\r\nAge: 0\r\nServer: ATS/10.0.0\r\n\r\nBody for response 6", '\000' <repeats 3917 times>
(gdb) c
Continuing.
Regression test(SDK_API_HttpAltInfo) still in progress

Breakpoint 1, synclient_txn_read_response (contp=<optimized out>) at traffic_server/InkAPITest.cc:669
669	  TSDebug(CDBG_TAG, "Response = |%s|, req len = %d", txn->response, txn->response_len);
(gdb) p txn->response
$4 = "HTTP/1.0 200 OK\r\nX-Response-ID: 6\r\nCache-Control: max-age=86400\r\nContent-Language: English\r\nDate: Tue, 08 Sep 2020 14:41:34 GMT\r\nAge: 0\r\nContent-Length: 19\r\nServer: ATS/10.0.0\r\n\r\nBody for response 6", '\000' <repeats 3897 times>
(gdb) c
Continuing.
[SDK_API_HttpAltInfo] TSHttpAltInfoClientReqGet : [TestCase] <<PASS>> { ok }
[SDK_API_HttpAltInfo] TSHttpAltInfoCachedReqGet : [TestCase] <<PASS>> { ok }
[SDK_API_HttpAltInfo] TSHttpAltInfoCachedRespGet : [TestCase] <<PASS>> { ok }
[SDK_API_HttpAltInfo] TSHttpAltInfoQualitySet : [TestCase] <<PASS>> { ok }

Breakpoint 1, synclient_txn_read_response (contp=<optimized out>) at traffic_server/InkAPITest.cc:669
669	  TSDebug(CDBG_TAG, "Response = |%s|, req len = %d", txn->response, txn->response_len);
(gdb) p txn->response
$5 = "HTTP/1.0 200 OK\r\nX-Response-ID: 6\r\nCache-Control: max-age=86400\r\nContent-Language: English\r\nDate: Tue, 08 Sep 2020 14:41:34 GMT\r\nAge: 60\r\nContent-Length: 19\r\nServer: ATS/10.0.0\r\n\r\nBody for response 6", '\000' <repeats 3896 times>
```

**Comment by ywkaras** [comment]
Converted this back to draft, as the tsapi test appears to be unstable.  Will be hard to debug, since it works consistently in my environment.

**Comment by ywkaras** [comment]
Includes fixes to problems in the TS core detected by the new tests.

**Comment by ywkaras** [comment]
Review hint:  most of the C++ code in tests/gold_tests/pluginTest/tsapi2 is an edited version of tests src/traffic_server/InkAPITests.cc that involve the "synthetic" client and server.

File: tests/gold_tests/pluginTest/tsapi/ts_tcp.h [hunk 2]
```diff
@@ -0,0 +1,1012 @@
...
+      auto *addr = reinterpret_cast<sockaddr_in const *>(TSNetVConnRemoteAddrGet(vconn));
+      TSReleaseAssert(addr != nullptr);
+      TSReleaseAssert(AF_INET == addr->sin_family);
+      TSReleaseAssert(htonl((static_cast<std::uint32_t>(127) << (3 * 8)) + 1) == addr->sin_addr.s_addr);
+      TSReleaseAssert(htons(_tcp_port_num) == addr->sin_port);
+
+      TSReleaseAssert(_conn_count < NUM_CONNECTIONS);
+
+      if (_server.get()->use_port_descriptor()) {
+        // Like the Regression Test for port descriptor, don't send any data, just close the connection.
```
Comment (ywkaras) [inline tests/gold_tests/pluginTest/tsapi/ts_tcp.h:838]: See https://github.com/apache/trafficserver/issues/6894 .

File: iocore/eventsystem/I_EventProcessor.h [hunk 3]
```diff
@@ -299,9 +299,9 @@ class EventProcessor : public Processor
   /// The thread group ID is the index into an array of these and so is not stored explicitly.
   struct ThreadGroupDescriptor {
     std::string _name;                               ///< Name for the thread group.
-    int _count                 = 0;                  ///< # of threads of this type.
-    std::atomic<int> _started  = 0;                  ///< # of started threads of this type.
-    uint64_t _next_round_robin = 0;                  ///< Index of thread to use for events assigned to this group.
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/I_EventProcessor.h:304]: Why was this changed from unsigned to signed?

**Comment by SolidWallOfCode** [comment]
Can you rebase this? There seems to be a lot of changes not related to testing (e.g., the changes in `ET_NET` thread tracking).

**Comment by ywkaras** [comment]
@duke8253 can you please review the changes by this PR to these files:
iocore/eventsystem/I_EventProcessor.h
iocore/eventsystem/P_UnixEventProcessor.h
iocore/net/UnixNetProcessor.cc
iocore/net/UnixNetVConnection.cc

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
[approve ci clang-format]

**Comment by bneradt** [comment]
[approve ci autest]

**Comment by bneradt** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
This PR at one time passed all the Au tests.  After rebasing to eliminate merge conflicts, it no longer does.  But there is no consensus that eliminating regression tests is a priority, so just making this a draft PR for now.
